src/services/epg.service.ts

```diff
@@ -0,0 +1,114 @@
+import { array, object, string } from 'yup';
+
+import type { PlaylistItem } from '#types/playlist';
+import { getDataOrThrow } from '#src/utils/api';
+import { isValidDateString } from '#src/utils/datetime';
+import { logDev } from '#src/utils/common';
+
+export const isFulfilled = <T>(input: PromiseSettledResult<T>): input is PromiseFulfilledResult<T> => input.status === 'fulfilled';
+
+export type EpgChannel = {
+  title: string;
+  image: string;
+  programs: EpgProgram[];
+};
+
+export type EpgProgram = {
+  id: string;
+  title: string;
+  startTime: string;
+  endTime: string;
+  description?: string;
+  image?: string;
+};
+
+const epgProgramSchema = object().shape({
+  id: string().required(),
+  title: string().required(),
+  startTime: string()
+    .required()
+    .test((value) => (value ? isValidDateString(value) : false)),
+  endTime: string()
+    .required()
+    .test((value) => (value ? isValidDateString(value) : false)),
+  chapterPointCustomProperties: array().of(
+    object().shape({
+      key: string().required(),
+      value: string().required(),
+    }),
+  ),
+});
+
+class EpgService {
+  async transformProgram(data: unknown) {
+    const incomingProgram = await epgProgramSchema.validate(data);
+    const transformed: EpgProgram = {
+      id: incomingProgram.id,
+      title: incomingProgram.title,
+      startTime: incomingProgram.startTime,
+      endTime: incomingProgram.endTime,
+    };
+
+    const customProperties = incomingProgram.chapterPointCustomProperties;
+
+    // map customProperties
+    if (customProperties) {
+      transformed.image = customProperties.find((item) => item.key === 'image')?.value || '';
+      transformed.description = customProperties.find((item) => item.key === 'description')?.value || '';
+    }
+
+    return transformed;
```

Comment:
```suggestion
    const program = await epgProgramSchema.validate(data);
    return {
      id: program.id,
      title: program.title,
      startTime: program.startTime,
      endTime: program.endTime,
      image: program.chapterPointCustomProperties?.find((item) => item.key === 'image')?.value || undefined,
      description: program.chapterPointCustomProperties?.find((item) => item.key === 'description')?.value || undefined
    };
```

Reply:
Simplifying a little to avoid having to get creative with local variable naming

Reply:
That's much better indeed!